docs/2024/puzzles/day15.md

```diff
@@ -2,10 +2,402 @@ import Solver from "../../../../../website/src/components/Solver.js"
+
+(With a mutable array, we’d have to be careful not to, say, overwrite an old box before it has been moved to its new position, and I don’t trust myself to reason that carefully. We could alternatively use a collection of objects with their positions instead of an array, which would make more sense for sparse grids, as finding adjacent objects would require traversing the entire collection.)
+
+Some preliminaries: enums to represent grid cells and move directions…
+
+```scala
+enum WarehouseCell:
+  case Empty
+  case Wall
+  case Box
+  case LeftBox
+  case RightBox
+  case Robot
+
+enum Direction:
+  case Up
+  case Down
+  case Left
+  case Right
+```
+
+…and a class to represent the state. In addition to the grid, we choose to (redundantly) track the position of the robot, since there’s only one and we might want its position often and don’t want to traverse the grid to find it every time.
+
+```scala
+class Warehouse(
+    val cells: ArraySeq[ArraySeq[WarehouseCell]],
+    val robotRow: Int,
+    val robotCol: Int)
+```
+
+*NB:* We use `ArraySeq`s instead of `Array`s to guarantee that we don’t accidentally mutate them in our program while still getting constant-time random access.
+
+### Part 1
+
+We start with the parser. First, we’ll need a way to create `Direction` and `WarehouseCell` objects from their `Char` representations. We could either write a simple pattern matcher:
+
+```scala
+object Direction:
+  def ofChar(c: Char) =
+    c match
+      case '^' => Up
+      case 'v' => Down
+      case '>' => Right
+      case '<' => Left
+```
+
+Or we could make the enum parametric, and use the parameter to define a `Map` (which we can automatically use as a `Char => WarehouseCell` function due to its `apply` method):
+
+```scala
+enum WarehouseCell(val repr: Char):
+  case Empty extends WarehouseCell('.')
+  case Wall extends WarehouseCell('#')
+  case Box extends WarehouseCell('O')
+  case LeftBox extends WarehouseCell('[')
+  case RightBox extends WarehouseCell(']')
+  case Robot extends WarehouseCell('@')
+
+object WarehouseCell:
+  val ofChar = Map.from(Seq(Wall, Box, Empty, Robot).map(c => (c.repr, c)))
+```
+
+As a personal challenge, I try to parse without reading the entire input into memory at once, only using the `File.getLines()` iterator and standard iterable methods.
+
+```scala
+object Day15:
+  def parseMoves(input: Iterator[String]): Seq[Direction] =
+    // extra flatMap because the moves may be split across multiple lines
+    input.flatMap(_.map(Direction.ofChar)).toSeq
+
+  def parse(inputFile: String): (Warehouse, Seq[Direction]) =
+    val file = io.Source.fromFile(inputFile)
```

Comment:
maybe `Using` instead of `try`/`finally`? I don't have a strong opinion here, just wondering if you'd considered it

Reply:
Is it more idiomatic? I'd have to look up how to use it :sweat_smile: but I could